File: android/feature-photos/src/main/kotlin/dev/omnesis/android/feature/photos/analyzers/SceneLabelAnalyzer.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.feature.photos.analyzers

import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.label.ImageLabeling
import com.google.mlkit.vision.label.defaults.ImageLabelerOptions
import dev.omnesis.android.feature.photos.PhotoAnalysisFragment
import dev.omnesis.android.feature.photos.PhotoAnalysisInput
import dev.omnesis.android.feature.photos.PhotoAnalyzer
import kotlinx.coroutines.tasks.await
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Scene/object labels via ML Kit's Image Labeling — new-photos tier only
 * (the backfill is OCR + place only). Folded into content AND promoted to
 * `metadata.tags` by `PhotosDocumentBuilder`, per the serving rule. Whole-
 * image classification, not localized detection — the direct parity match
 * to iOS's `SceneLabelAnalyzer` (`VNClassifyImageRequest`, also
 * whole-image-only); ML Kit's separate Object Detection & Tracking API is
 * built for localized bounding-box detection across video frames and would
 * over-shoot this need.
 */
class SceneLabelAnalyzer(
    private val confidenceThreshold: Float = 0.3f,
    private val maxLabels: Int = 5,
) : PhotoAnalyzer {
    override val identifier = "scene-labels"

    private val labeler = ImageLabeling.getClient(
        ImageLabelerOptions.Builder().setConfidenceThreshold(confidenceThreshold).build(),
    )

    override suspend fun isAvailable(): Boolean = true

    override suspend fun analyze(input: PhotoAnalysisInput): PhotoAnalysisFragment? {
        val bitmap = input.bitmap ?: return null
        val labels = labeler.process(InputImage.fromBitmap(bitmap, 0)).await()
            .sortedByDescending { it.confidence }
            .take(maxLabels)
        if (labels.isEmpty()) return null

        val confidences = labels.associate { it.text to JsonPrimitive(it.confidence.toDouble()) }
        return PhotoAnalysisFragment(
            tags = labels.map { it.text },
            extra = mapOf("sceneLabelConfidence" to JsonObject(confidences)),
        )
    }
}

File: ios/Sources/Omnesis/Photos/Analyzers/SceneLabelAnalyzer.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(Vision)
import Vision

/// Scene/object labels via `VNClassifyImageRequest` — new-photos tier
/// only (the issue scopes rich analysis to new photos; the historical
/// backfill is OCR + place + metadata only). Folded into `content` AND
/// promoted to `metadata.tags` by `PhotosDocumentBuilder`, per the
/// serving rule (a label reachable only via `extra` is never found by
/// search).
public struct SceneLabelAnalyzer: PhotoAnalyzer {
    public let identifier = "scene-labels"

    /// Only labels at or above this confidence are kept — Vision's
    /// classifier returns hundreds of candidates per image, most far
    /// too weak to be a meaningful searchable tag.
    private let confidenceThreshold: Float
    /// Cap on how many labels fold into tags/content per photo.
    private let maxLabels: Int

    public init(confidenceThreshold: Float = 0.3, maxLabels: Int = 5) {
        self.confidenceThreshold = confidenceThreshold
        self.maxLabels = maxLabels
    }

    public func isAvailable() async -> Bool {
        true
    }

    public func analyze(_ input: PhotoAnalysisInput) async -> PhotoAnalysisFragment? {
        guard let image = input.image else { return nil }
        let request = VNClassifyImageRequest()
        let handler = VNImageRequestHandler(cgImage: image, options: [:])
        do {
            try handler.perform([request])
        } catch {
            return nil
        }
        let labels = (request.results ?? [])
            .filter { $0.confidence >= confidenceThreshold }
            .sorted { $0.confidence > $1.confidence }
            .prefix(maxLabels)
        guard !labels.isEmpty else { return nil }

        var extra: [String: JSONValue] = [:]
        for label in labels {
            extra[label.identifier] = .double(Double(label.confidence))
        }
        return PhotoAnalysisFragment(
            tags: labels.map(\.identifier),
            extra: ["sceneLabelConfidence": .object(extra)]
        )
    }
}
#endif
